### PR TITLE
Headless Blazor client

### PR DESCRIPTION
**Pull request by NTaylorMullen**
@pranavkm / @javiercn this should at least enable you to start a client and connect to a server. I did the first pass on building a rudimentary object model to represent the render batch so I can properly invoke events etc. but i'm seeing some wonkyness with what I have currently. Will sync up with @SteveSandersonMS on what i'm missing here tomorrow.

An example of the serialized model (yes it's buggy)
```
[Component ( 0 )]
[Component ( 1 )]
    M: \r\n    
    M: \r\n
[Component ( 2 )]
[Component ( 3 )]
[Component ( 4 )]
    <div class="sidebar">
        M: \r\n    
        M: \r\n
    </div/>
    M: \r\n\r\n
    <div class="main">
        M: \r\n\r\n    
        <div class="content px-4">
            M: \r\n        
            M: \r\n    
        </div/>
        M: \r\n
    </div/>
[Component ( 5 )]
    <div class="top-row pl-4 navbar navbar-dark">
        M: \r\n    
        M: <a class="navbar-brand" href>Razor Components</a>\r\n    
        <button class="navbar-toggler"  Events: [ click(1)]>
            M: \r\n        <span class="navbar-toggler-icon"></span>\r\n    
        </button/>
        M: \r\n
    </div/>
    M: \r\n\r\n
    <div class="collapse"  Events: [ click(2)]>
        M: \r\n    
        <ul class="nav flex-column">
            M: \r\n        
            <li class="nav-item px-3">
                M: \r\n            
                M: \r\n        
            </li/>
            M: \r\n        
            <li class="nav-item px-3">
                M: \r\n            
                M: \r\n        
            </li/>
            M: \r\n        
            <li class="nav-item px-3">
                M: \r\n            
                M: \r\n        
            </li/>
            M: \r\n        
            <li class="nav-item px-3">
                M: \r\n            
                M: \r\n        
            </li/>
            M: \r\n        
            <li class="nav-item px-3">
                M: \r\n            
                M: \r\n        
            </li/>
            M: \r\n        
            <li class="nav-item px-3">
                M: \r\n            
                M: \r\n        
            </li/>
            M: \r\n    
        </ul/>
        M: \r\n
    </div/>
[Component ( 6 )]
    M: <h1>Counter</h1>\r\n\r\n
    <p>
        T: Current count: 
        T: 0
    </p/>
    M: \r\n\r\n
    <button class="btn btn-primary" id="thecounter"  Events: [ click(3)]>
        T: Click me
    </button/>
[Component ( 7 )]
    <a class="nav-link" href="" Match="All">
    </a/>
[Component ( 8 )]
    <a class="nav-link active" href="counter">
    </a/>
[Component ( 9 )]
    <a class="nav-link" href="fetchdata">
    </a/>
[Component ( 10 )]
    <a class="nav-link" href="ticker">
    </a/>
[Component ( 11 )]
    <a class="nav-link" href="greeter">
    </a/>
[Component ( 12 )]
    <a class="nav-link" href="error">
    </a/>

```

For page:
```
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width">
    <title>Razor Components</title>
    <base href="/" />
    <link href="css/bootstrap/bootstrap.min.css" rel="stylesheet" />
    <link href="css/site.css" rel="stylesheet" />
</head>
<body>
    <app><!-- M.A.C.Component:{"circuitId":"CfDJ8FT_U1j9m7BBifX121yqb5llIUwY_uSS1dAuTDv9GNYG8q8WtcHzp41J5wQ89t-BmKylO3rPkjjt-RTrE7rfYmkVlQeHkjHmXzgIliLdhz5CDmJ4sTBoV85bao7TpIKr9lDnjylpa0Y5w4Kwdjit4CnRaosIb1GTzNHPwNmB1Olc","rendererId":"0","componentId":"0"} -->
    <div class="sidebar">
    <div class="top-row pl-4 navbar navbar-dark">
    <a class="navbar-brand" href>Razor Components</a>
    <button class="navbar-toggler">
        <span class="navbar-toggler-icon"></span>
    </button>
</div>

<div class="collapse">
    <ul class="nav flex-column">
        <li class="nav-item px-3">
            <a class="nav-link" href="" Match="All">
                <span class="oi oi-home" aria-hidden="true"></span> Home
            </a>
        </li>
        <li class="nav-item px-3">
            <a class="nav-link active" href="counter">
                <span class="oi oi-plus" aria-hidden="true"></span> Counter
            </a>
        </li>
        <li class="nav-item px-3">
            <a class="nav-link" href="fetchdata">
                <span class="oi oi-list-rich" aria-hidden="true"></span> Fetch data
            </a>
        </li>
        <li class="nav-item px-3">
            <a class="nav-link" href="ticker">
                <span class="oi oi-list-rich" aria-hidden="true"></span> Ticker
            </a>
        </li>
        <li class="nav-item px-3">
            <a class="nav-link" href="greeter">
                <span class="oi oi-list-rich" aria-hidden="true"></span> Greeter
            </a>
        </li>
        <li class="nav-item px-3">
            <a class="nav-link" href="error">
                <span class="oi oi-list-rich" aria-hidden="true"></span> Error
            </a>
        </li>
    </ul>
</div>
</div>

<div class="main">

    <div class="content px-4">
        <h1>Counter</h1>

<p>Current count: 0</p>

<button class="btn btn-primary" id="thecounter">Click me</button>
    </div>
</div>
<!-- M.A.C.Component: 0 --></app>

    <script src="_framework/blazor.server.js" autostart="false"></script>
    <script>
        Blazor.start({
            configureSignalR: function (builder) {
                builder.configureLogging(2); // LogLevel.Information
            },
            logLevel: 2 // LogLevel.Information
        }).then(function () {
            window['__aspnetcore__testing__blazor__started__'] = true;
        });
    </script>
</body>
</html>

```

### CODE REVIEW

File: src/Components/test/Ignitor/ContainerNode.cs
Comment (pranavkm) [inline src/Components/test/Ignitor/ContainerNode.cs]: Formatting

File: src/Components/test/Ignitor/ElementHive.cs
Comment (pranavkm) [inline src/Components/test/Ignitor/ElementHive.cs]: Casing

File: src/Components/test/Ignitor/ElementHive.cs
Comment (pranavkm) [inline src/Components/test/Ignitor/ElementHive.cs]: ~~this.~~

File: src/Components/test/Ignitor/Program.cs
Comment (pranavkm) [inline src/Components/test/Ignitor/Program.cs]: Sort usings

File: src/Components/test/Ignitor/Program.cs
Comment (pranavkm) [inline src/Components/test/Ignitor/Program.cs]: I imagine we'd have to do something more here eventually to test out JSInterop scenarios, but this is fine for a start

File: src/Components/test/Ignitor/RenderBatchReader.cs
Comment (pranavkm) [inline src/Components/test/Ignitor/RenderBatchReader.cs]: I think some of these might have benefited from using a `SequenceReader`. That way you wouldn't have to manually track the indices. But I wouldn't bother changing the codez at this point.

File: src/Components/test/Ignitor/RenderBatchReader.cs
Comment (pranavkm) [inline src/Components/test/Ignitor/RenderBatchReader.cs]: Like with the `SequenceReader`, this would basically be `new Sections(reader.ReadInt32(), reader.ReadInt32(), ...)`. But I may be overlooking something.

**Review by pranavkm — APPROVED** [review]

File: src/Components/Components/src/RenderTree/RenderTreeFrame.cs
Comment (pranavkm) [reply src/Components/Components/src/RenderTree/RenderTreeFrame.cs]: Also FYI @SteveSandersonMS

**Comment by NTaylorMullen** [comment]
Started adding click/event handler support.

File: src/Components/Components.sln
Comment (SteveSandersonMS) [inline src/Components/Components.sln]: Do these three need to be added to the solution, or is this just a temporary thing while you're working on this PR?

File: src/Components/Components.sln
Comment (SteveSandersonMS) [reply src/Components/Components.sln]: Also, should the name be something like `Microsoft.AspNetCore.Components.Test.Ignitor`? I know that's pretty long though!

File: src/Components/Components/src/RenderTree/RenderTreeFrame.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/RenderTreeFrame.cs]: I'd like to understand what the change is required for. Before this change, it was impossible to have a nonempty `ComponentId` without also having a `ComponentState`, but this change would allow that situation to occur. Why do we want that to be possible?

None of the static factory methods that call this constructor make use of this possibility. Am I missing the reason why this extra param is required?

File: src/Components/test/Ignitor/CommentNode.cs
Comment (SteveSandersonMS) [inline src/Components/test/Ignitor/CommentNode.cs]: Is this called `CommentNode` because it plays the same role as the logical element containers in the TypeScript code? The TypeScript code happens to use DOM comment nodes as an implementation detail, but there's no reason it would have to be comment nodes if some other node type happened to be more suitable.

If the purpose of `CommentNode` here is to act as a logical element container, I suggest renaming it to `LogicalContainerNode` or something like that, rather than implying it needs to be a "comment".

File: src/Components/test/Ignitor/ElementHive.cs
Comment (SteveSandersonMS) [inline src/Components/test/Ignitor/ElementHive.cs]: I don't know how clean and future-proof we want to make this code (depends on our intentions around its long-term use). But if we have time, it would be interesting to see how much it can be decoupled from HTML concepts.

For example, we could have a base `ElementHive` with the majority of the logic here plus a few virtual methods. A subclass, `HtmlElementHive`, would layer on specific knowledge of HTML behaviors such as what's here.

The reason for this is that we're trying to keep the core Components library agnostic to web-specific concepts, to leave open as much flexibility as possible for it to be used with non-web rendering technologies in the future. That's not to say we're committed to one day shipping support for a non-web rendering technology, but the easier we make it for ourselves, the more capable we'd be of experimenting with it (or maybe others would).

File: src/Components/test/Ignitor/ElementNode.cs
Comment (SteveSandersonMS) [inline src/Components/test/Ignitor/ElementNode.cs]: Is it really useful to distinguish between attributes and properties in this test infrastructure? What use will we make of that distinction?

Comment (SteveSandersonMS) [reply src/Components/test/Ignitor/ElementNode.cs]: I'm not dead set against doing it as you have here, though. Just opening the question about what our goals are.

File: src/Components/test/Ignitor/Ignitor.csproj
Comment (SteveSandersonMS) [inline src/Components/test/Ignitor/Ignitor.csproj]: AngleSharp?? Is this because you're planning to parse markup frame content at some point?

File: src/Components/test/Ignitor/RenderBatchReader.cs
Comment (SteveSandersonMS) [inline src/Components/test/Ignitor/RenderBatchReader.cs]: We have a set of unit tests that cover our `RenderBatchWriter`. If you want to have unit test coverage for `RenderBatchReader` as well, you could go to those unit tests and append some assertions that the output from the writer can be fed back into your reader and it recovers the original data.

File: src/Components/test/Ignitor/Program.cs
Comment (BrennanConroy) [inline src/Components/test/Ignitor/Program.cs]:
```suggestion
            await using var connection = builder.Build();
```

File: src/Components/Components/src/RenderTree/RenderTreeFrame.cs
Comment (NTaylorMullen) [reply src/Components/Components/src/RenderTree/RenderTreeFrame.cs]: We'll have to ask @rynowak his thoughts on this one. I've undone it locally for now and all seems to work.

File: src/Components/Components.sln
Comment (NTaylorMullen) [reply src/Components/Components.sln]: Ehh, thinking I should probably just move this into the testassets folder instead and leave as Ignitor to align with the other test apps we have there.

File: src/Components/Components.sln
Comment (NTaylorMullen) [reply src/Components/Components.sln]: They don't have to in the filtered sln but in this one they need to be here.

File: src/Components/test/Ignitor/ElementNode.cs
Comment (NTaylorMullen) [reply src/Components/test/Ignitor/ElementNode.cs]: Definitely, for this first pass I let the browser renderer drive the requirements of the intermediate representation (of which it set properties). If we find it's not useful we can nuke that later. This iteration is trying to work consistently with what the HTML browser renderer does.

File: src/Components/test/Ignitor/Ignitor.csproj
Comment (NTaylorMullen) [reply src/Components/test/Ignitor/Ignitor.csproj]: lmao, def not. Going to remove that

File: src/Components/test/Ignitor/RenderBatchReader.cs
Comment (NTaylorMullen) [reply src/Components/test/Ignitor/RenderBatchReader.cs]: > We have a set of unit tests that cover our `RenderBatchWriter`. If you want to have unit test coverage for `RenderBatchReader` as well, you could go to those unit tests and append some assertions that the output from the writer can be fed back into your reader and it recovers the original data.

Great suggestion.

> Assuming we're going to keep and maintain this code, I think it would be very valuable to have unit test coverage for it. We're effectively duplicating a bunch of knowledge here, so whenever we add or change to the format of frames/edits or the serialization format, we have to remember to come and make the equivalent changes. Things could easily drift apart without sufficiently detailed checks.

Ya I totally agree. I'm going to take a quick look at those unit tests and see if it's easy to bring over. Going to time box that to 30 minutes, if it takes more we'll have to revisit so I can get all my preview7 work done 😄

File: src/Components/test/Ignitor/Program.cs
Comment (NTaylorMullen) [reply src/Components/test/Ignitor/Program.cs]: Pranav points!!!

File: src/Components/test/testassets/ComponentsApp.App/Pages/Counter.razor
Comment (NTaylorMullen) [inline src/Components/test/testassets/ComponentsApp.App/Pages/Counter.razor]: This enables us to lookup this button in the headless client to click it.

File: src/Components/test/testassets/Ignitor/Program.cs
Comment (NTaylorMullen) [inline src/Components/test/testassets/Ignitor/Program.cs]: @javiercn / @pranavkm this causes rendering to re-render infinitely. It illustrates how to click though 😄

File: src/Components/test/Ignitor/RenderBatchReader.cs
Comment (NTaylorMullen) [reply src/Components/test/Ignitor/RenderBatchReader.cs]: Left as-is for now.

File: src/Components/test/Ignitor/ElementHive.cs
Comment (NTaylorMullen) [reply src/Components/test/Ignitor/ElementHive.cs]: Talked offline, this first pass at this tool is mostly to unblock Javier/Pranav. We can revisit later.

**Comment by NTaylorMullen** [comment]
Alright, once tests pass going to merge this.

@SteveSandersonMS I was able to consume a lot of the `RenderBatchWriter` tests but didn't add the disposed event handlers and component tests (2 of them) because we don't handle those in Ignitor.

File: src/Components/test/testassets/Ignitor/Program.cs
Comment (NTaylorMullen) [reply src/Components/test/testassets/Ignitor/Program.cs]: So to write your perf tests you'll have to modify this `Program.cs` (which takes a URL) and then if you want to see the model that's generated on the hive you can inspect its `SerializedValue`.

**Comment by NTaylorMullen** [comment]
@aspnet/build I imagine this PR is experiencing the hang bug that @natemcmaster notified us all about? What's "the right thing" to do next to enable me to get this PR in?

Do I have to re-run all the checks? 😢

**Comment by dougbu** [comment]
This is actually a slightly different problem that's occurring less often than hanging jobs. AzDO is sometimes not reporting the final status of builds back to GitHub.

In this case, the build was successful modulo a flaky test or two. I'll merge.

**Comment by dougbu** [comment]
Oh, is this PR actually a WIP still?

**Comment by natemcmaster** [comment]
Looks like AzDO regressed again on reporting final PR check status. I see https://github.com/dotnet/core-eng/issues/6416 was re-opened 27 minutes ago.

**Comment by NTaylorMullen** [comment]
> Oh, is this PR actually a WIP still?

Nope, no longer WIP, feel free to merge 😄 

> Looks like AzDO regressed again on reporting final PR check status. I see [dotnet/core-eng#6416](https://github.com/dotnet/core-eng/issues/6416) was re-opened 27 minutes ago.

😢

**Comment by javiercn** [comment]
@NTaylorMullen I've been looking at this. How can I use the client to test arbitrary logic? Is your expectation that I write my own Program.cs and simply reference ignitor as a library?

**Comment by javiercn** [comment]
I've added a buch of stuff here https://github.com/aspnet/AspNetCore/pull/11289

**Comment by NTaylorMullen** [comment]
> How can I use the client to test arbitrary logic? Is your expectation that I write my own Program.cs and simply reference ignitor as a library?

That's what I wanted to do but wasn't required for the initial iteration. Basically it'd be nice to have it just be a class library with enough API to host it as a console app or w/e.

To unblock everyone the intent was to just change the Ignitor client in-place.

> I've added a buch of stuff here #11289

Yup, that's a nice middle ground to what I want to get at eventually.

**Pull request merged**
